### ISSUE TITLE
Automatically blacklist workers who auto-accept HITs but don't start them

### ISSUE THREAD

**Issue by thomasmorgan** [Dallinger/Dallinger#2381]
Some workers use auto-accepting scripts to quickly accept HITs as they are published. However, this typically adds the HIT to the workers queue and there can be a significant delay before they start the work. This is fine in some cases (and these workers do good work) but it causes problems when experiments rely on groups of participants completing the work synchronously as groups can take a long time to fill.

The current solution is to over-recruit by as much as 50%, thus there is excess recruitment space that the auto-accepting workers can fill, while other participants can take part. However, this is inefficient for two reasons: (1) it is hard to know how much to over-recruit by, and (2) over-recruitment typically leads to excess participant fees as participants who arrive after all networks are full are still paid for trying to take part.

I suggest the following solution:
Dallinger automatically detects auto-accepting participants, and gives them a hidden qualification that optionally blocks them from future HITs put up by the experimenter.

FAQ:
How does Dallinger detect auto-accepting participants?
A background process can monitor for any participants who arrive >10 minutes after their assignment accepted notification arrived.

How does Dallinger assign the qualification?
Dallinger can already do this.

How does the qualification block participants?
It can be automatically added to the `qualification_blacklist` variable at runtime

How is the blocking made optional?
The config file can have a boolean parameter that determines whether the qualification is added to the `qualification_blacklist` variable at runtime, e.g. `block_autoaccepting_workers = true`

What about false positives?
We can give each participant "three lives", and each time they don't start working within 10 minutes the value of their qualification increases by 1. We can then block only those workers who have the qualification with a value of 3 or more.

@jacobyn @lottybrand @jessesnyder What do you think?

**Comment by pmcharrison** [comment]
Hi @thomasmorgan , I thought that setting a short HIT duration was the solution to this - it means that people can't queue up HITs because they will just expire before they get a chance to do them. But did you find that this doesn't work in practice? I guess there is no penalty for accepting a HIT and letting it expire, so people might just do that.

**Comment by thomasmorgan** [comment]
Thanks @pmcharrison. That works fine, except in synchronous group experiments where the auto-accepting works hold up their entire group (the group cannot start until everyone has arrived) which causes everyone to expire, even those who have been patiently waiting the whole time.

**Comment by lottybrand** [comment]
Thanks Tom. A few things to add here: 
1) It's not just inefficient and costly because of the reasons you lay out, but it also means recruiting an extra one or two groups at a time if needed becomes very costly/almost impossible, and you can lose data from the final groups to fill, because:
a) if there are only two groups to fill, these quickly get filled by auto-accepters so the time it takes for them to start means often some participants are sat there for longer than the experiment is set to last, so in anticipation of the timer running out on them they either panic and leave, or get annoyed and leave, or they get kicked out as the timer actually runs out
b) the same thing happens for the final one/two groups as above, meaning sometimes you lose those entire groups. 
c) you pay considerably more for over-recruiting less participants than more participants, sometimes double-fees
d) you have to be "on the ball" watching and waiting for the groups to fill so you can expire the HIT as soon as they fill to avoid overpaying for over-recruiters

I think your solution would help, but, it would still mean we have the above problems for the first few runs? Or you mean Dallinger collects this autoaccept data on all future experiments, not just group experiments? 

The only other solutions I can think of is either just booking in certain times for participants to take part with a qualification HIT first, or, Dallinger somehow kicks out the autorecruiters live during the group experiment? eg if the timer finds they have been more than 5-10 minutes on the instruction page then it kicks them out there and then...? We would have to put this warning in the advert for the HIT though

**Comment by thomasmorgan** [comment]
Thanks @lottybrand. Yes, this would take a few runs to kick in, but it would then apply to all future experiments, so it wouldn't reset for each new experiment.

Kicking auto-accepters out in real time would be the dream, but I don't think MTurk let's us "force return" HITs on workers behalves. Booking certain times for qualified workers sounds like more trouble than it is worth IMO, but it could be done.

**Comment by pmcharrison** [comment]
> I don't think MTurk let's us "force return" HITs on workers

Couldn't Dallinger simply call an AssignmentSubmitted event if the participant didn't start working within 2 minutes? so, it's like the participant completed the experiment and submitted the HIT? just without getting a bonus? wouldn't that be a way of "force returning"?

**Comment by thomasmorgan** [comment]
@pmcharrison yes that would work, though I worry about the incentive structure it sets up: we would reward (i.e. pay) participants for doing nothing. To avoid this we could set the base pay really low and have everything ride on the bonus. But this would put regular workers off too.

**Comment by pmcharrison** [comment]
Yep, this is actually what we do currently, have an essentially zero base payment, and only pay people with the bonus. It works well for us, actually.

**Comment by thomasmorgan** [comment]
@pmcharrison How do you handle participants who glitch out during the experiment. My usual policy is to push them to the debriefing and give them the base payment, but they typically don't get a bonus because their data is bad (and so a bonus can't necessarily be calculated). I guess the solution is to have the bonus function bottom-out at a decent value, so everyone who submits gets some bonus, but only those who make it to the end get bigger bonuses.

**Comment by pmcharrison** [comment]
In participants glitching out, we don't rely on a 'graceful' catch that takes participants to debriefing - in our cases, if a glitch happens, it usually means their session is not retrievable. So, we have them email us, and we manually compensate them according to their story / how far they had progressed in the experiment. For this we use the new Dallinger compensation command, which speeds things up a lot.

**Comment by thomasmorgan** [comment]
Ah ok, I've been running a lot of grouped experiments which have quite high attrition rates, so the graceful catching is really useful. Still, it would be pretty easy to give everyone who finishes a bonus even if their data is garbled.
